### DISCUSSION TITLE
Alternative to buffer protocol

### DISCUSSION THREAD

**Discussion by awni** [wjakob/nanobind#478] [Q&A] [answered]
We are evaluating switching MLX from Pybind11 to Nanobind (https://github.com/ml-explore/mlx/pull/839)

The `nb::ndarray` seems to be fine for constructing MLX arrays from e.g. incoming NumPy arrays. However, we use the buffer protocol to be able to do:

```
np.array(mlx_array)
```

However, nanobind does not have the same support for the buffer protocol. I am wondering if there is a recommended approach to achieve the same as above?

**Comment by wjakob** [comment, marked as answer]
Please see this part of the documentation: https://nanobind.readthedocs.io/en/latest/typeslots.html

You should be able to add custom slots recreate the pybind11 behavior if you truly want your object to expose the native buffer protocol. What I think would be much easier is to add an ``__array__`` method to your class, which NumPy will accept as well.

**Reply by awni** [reply]
Yes I considered `__array__` which works fine, but just for numpy. Instead I just reimplemented the buffer protocol as you suggested. It is not too bad actually.
